### PR TITLE
Add question mark notice lint

### PR DESCRIPTION
**Pull request by kumarde**
Returns a `Notice` if the commonName or SAN DNSNames has an item that is prepended with `?.`. 

Addresses https://github.com/zmap/zlint/issues/165

### CODE REVIEW

File: lints/lint_dnsname_contains_question_marks.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// lint_dnsname_contains_question_marks.go
```
Comment (dadrian) [inline lints/lint_dnsname_contains_question_marks.go]: Delete this line, add a copyright notice.

Comment (kumarde) [reply lints/lint_dnsname_contains_question_marks.go]: Copyright?

File: lints/lint_dnsname_contains_question_marks.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+			return &LintResult{Status: Notice}
+		}
+	}
+	return &LintResult{Status: Pass}
+}
+
+func init() {
+	RegisterLint(&Lint{
+		Name:          "n_dnsname_contains_question_marks",
+		Description:   "Some Precerts are prepended with question marks.",
```
Comment (zakird) [inline lints/lint_dnsname_contains_question_marks.go]: This has a term---"redacted". I'd make the name and description "n_contains_redacted_dnsname" and the source to be ZLint.

File: lints/lint_dnsname_contains_question_marks.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+func isPrependedByQuestionMarks(domain string) bool {
+	if strings.HasPrefix(domain, "?.") {
+		return true
+	}
+	return false
+}
+
+func (l *DNSNameContainsQuestionMarks) Execute(c *x509.Certificate) *LintResult {
```
Comment (zakird) [inline lints/lint_dnsname_contains_question_marks.go]: This should probably only run if `cn != ""`

Comment (zakird) [reply lints/lint_dnsname_contains_question_marks.go]: This really only checks if this is prepended by _a_ question mark, not multiple question marks.

File: lints/lint_dnsname_contains_question_marks.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	"strings"
+)
+
+type DNSNameContainsQuestionMarks struct{}
+
+func (l *DNSNameContainsQuestionMarks) Initialize() error {
+	return nil
+}
+
+func (l *DNSNameContainsQuestionMarks) CheckApplies(c *x509.Certificate) bool {
```
Comment (zakird) [inline lints/lint_dnsname_contains_question_marks.go]: This probably only applies to subscriber certificates.

File: lints/lint_dnsname_contains_question_marks.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	return true
+}
+
+func isPrependedByQuestionMarks(domain string) bool {
+	if strings.HasPrefix(domain, "?.") {
```
Comment (zakird) [inline lints/lint_dnsname_contains_question_marks.go]: This function returns a bool...

File: util/fqdn.go [hunk 6]
```diff
@@ -7,15 +7,15 @@ import (
 	"strings"
 )
 
-func removeQuestionMarks(domain string) string {
+func RemoveQuestionMarks(domain string) string {
```
Comment (zakird) [inline util/fqdn.go]: This doesn't look like this handles multiple question marks. I think `?.?.domain` is an acceptable redacted certificate.

Comment (kumarde) [reply util/fqdn.go]: it does - it's in a for loop.

File: lints/lint_dnsname_contains_question_marks.go (hunk 2 above)
Comment (kumarde) [reply lints/lint_dnsname_contains_question_marks.go]: done

File: lints/lint_dnsname_contains_question_marks.go (hunk 4 above)
Comment (kumarde) [reply lints/lint_dnsname_contains_question_marks.go]: done

File: lints/lint_dnsname_contains_question_marks.go (hunk 5 above)
Comment (kumarde) [reply lints/lint_dnsname_contains_question_marks.go]: done

File: util/fqdn.go (hunk 6 above)
Comment (zakird) [reply util/fqdn.go]: Then it should probably be named `removePrependedQuestionMark`. The function name implies that it removes multiple and that's not the case.

File: util/fqdn.go [hunk 7]
```diff
@@ -7,15 +7,15 @@ import (
...
 	for strings.HasPrefix(domain, "?.") {
 		domain = domain[2:]
 	}
 	return domain
 }
 
 func IsFQDN(domain string) bool {
-	domain = removeQuestionMarks(domain)
+	domain = RemoveQuestionMarks(domain)
 	if strings.HasPrefix(domain, "*.") {
```
Comment (zakird) [inline util/fqdn.go]: Why does one of these get a helper function when the second one doesn't? They have identical logic.

File: lints/lint_dnsname_contains_question_marks.go (hunk 3 above)
Comment (kumarde) [reply lints/lint_dnsname_contains_question_marks.go]: Yeah, I think that encompasses everything we want to check though? `?.?.abc.com` is still prepended with `?.` Maybe I'm not understanding this comment

Comment (zakird) [reply lints/lint_dnsname_contains_question_marks.go]: The thing should really just be named "IsRedactedCertificate"

File: util/fqdn.go (hunk 7 above)
Comment (kumarde) [reply util/fqdn.go]: They don't - the first runs in a for loop. I can add a helper function if you'd like, but don't think we necessarily need it.

Comment (zakird) [reply util/fqdn.go]:
```func isPrependedByQuestionMarks(domain string) bool {
	if strings.HasPrefix(domain, "?.") {
		return true
	}
	return false
}```

Comment (zakird) [reply util/fqdn.go]: ^^ That has the exact same logic there.

**Comment by titanous** [comment]
`*.?` is also used: https://crt.sh/?q=*.%3F.%25

**Comment by kumarde** [comment]
@zakird @titanous fixed up those comments.

File: util/fqdn.go [hunk 8]
```diff
@@ -7,18 +7,23 @@ import (
...
+func RemovePrependedQuestionMarks(domain string) string {
 	for strings.HasPrefix(domain, "?.") {
 		domain = domain[2:]
 	}
 	return domain
 }
 
-func IsFQDN(domain string) bool {
-	domain = removeQuestionMarks(domain)
+func RemoveWildcardFromDomain(domain string) string {
```
Comment (zakird) [inline util/fqdn.go]: Can you name these two functions consistently?

File: lints/lint_dnsname_bad_character_in_label.go [hunk 9]
```diff
@@ -33,13 +33,15 @@ func (l *DNSNameProperCharacters) labelContainsBadCharacters(domain string) bool
 
 func (l *DNSNameProperCharacters) Execute(c *x509.Certificate) *LintResult {
 	if c.Subject.CommonName != "" {
```
Comment (zakird) [inline lints/lint_dnsname_bad_character_in_label.go]: It doesn't look like this correctly handles `*.?.bad.com`

Comment (kumarde) [reply lints/lint_dnsname_bad_character_in_label.go]: AH, good catch.

File: util/fqdn.go (hunk 8 above)
Comment (kumarde) [reply util/fqdn.go]: done

File: lints/lint_dnsname_bad_character_in_label.go (hunk 9 above)
Comment (kumarde) [reply lints/lint_dnsname_bad_character_in_label.go]: i fixed that btw.

File: lints/lint_dnsname_contains_question_marks.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+	}
+	return &LintResult{Status: Pass}
+}
+
+func init() {
+	RegisterLint(&Lint{
+		Name:          "n_contains_redacted_dnsname",
+		Description:   "Some Precerts are prepended with question marks.",
+		Source:        CABFBaselineRequirements,
+		Citation:      "MDSP",
```
Comment (zakird) [inline lints/lint_dnsname_contains_question_marks.go]: This citation doesn't make sense.

File: lints/lint_dnsname_contains_question_marks.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
+package lints
```
Comment (zakird) [inline lints/lint_dnsname_contains_question_marks.go]: This file needs to be renamed.

File: lints/lint_is_redacted_cert.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+			return &LintResult{Status: Notice}
+		}
+	}
+	return &LintResult{Status: Pass}
+}
+
+func init() {
+	RegisterLint(&Lint{
+		Name:          "n_contains_redacted_dnsname",
+		Description:   "Some Precerts are prepended with question marks.",
```
Comment (zakird) [inline lints/lint_is_redacted_cert.go]: This description, source, and citation don't make much sense.

**Pull request merged**